sled-agent/src/common/instance.rs

```diff
@@ -254,6 +260,48 @@ impl InstanceStates {
         }
     }
 
+    fn request_migrating(
+        &mut self,
+        migration_id: Option<Uuid>,
+    ) -> Result<Option<Action>, Error> {
+        let migration_id = migration_id.ok_or_else(|| {
+            Error::invalid_request(
+                "expected migration id to transition to \"migrating\" state",
+            )
+        })?;
+        match self.current.run_state {
+            // Valid states for a migration request
+            InstanceState::Running => {
+                self.current.migration_uuid = Some(migration_id);
+                self.transition(
+                    InstanceState::Migrating,
+                    Some(InstanceStateRequested::Running),
+                );
+                // We don't return any action here for propolis to take
+                // because the actual migration process will be driven
+                // via a followup request to sled agent.
+                return Ok(None);
+            }
+            // Invalid states for a migration request
+            InstanceState::Creating
+            | InstanceState::Starting
+            | InstanceState::Stopping
+            | InstanceState::Stopped
+            | InstanceState::Rebooting
+            | InstanceState::Migrating
```

Comment:
If we're already migrating, and the same migration ID is requested, I think it's important that we return "OK" with no action here. Without that, I think the saga all the way up in Nexus is not idempotent, which breaks the (necessary) contract of sagas.

Here's the complete picture:

We call the saga step, `sim_migrate_prep` (which is where the idempotency requirement starts):
https://github.com/oxidecomputer/omicron/blob/5b2d4a5c4b192fb4e1bb5499f01510ee0d235142/nexus/src/sagas.rs#L382-L402

This calls `instance_start_migrate`, which requests that we Migrate with a particular ID by calling `instance_set_runtime`...
https://github.com/oxidecomputer/omicron/blob/5b2d4a5c4b192fb4e1bb5499f01510ee0d235142/nexus/src/nexus.rs#L1206-L1214

This tells the Sled Agent to perform the migration, and write the updated state to the DB...
https://github.com/oxidecomputer/omicron/blob/5b2d4a5c4b192fb4e1bb5499f01510ee0d235142/nexus/src/nexus.rs#L1248-L1265

*However*, it's totally possible we make the call to the Sled Agent - requesting that the migration starts - and before we get a chance to update the DB (by calling `db_datastore.instance_update_runtime`), Nexus crashes.

If this happens, the saga action would get restarted on recovery, and we'd go through all those same steps.

This would start by trying to `instance_put` a sled with a `Migrating` `run_state`, and a particular `migration_id`. However, when we get to this point within the Sled Agent, that call would now fail, because the migration already started!

Reply:
Good point, tweaked this to make sure it is idempotent in that scenario as well.